File: LeetCode/src/graph/topologicsort/CourseScheduleII.java
package graph.topologicsort;

import java.util.*;

public class CourseScheduleII {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] indegrees = new int[numCourses];
        List<List<Integer>> graph = new ArrayList<>();

        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] pre : prerequisites) {
            int first = pre[1];
            int second = pre[0];
            graph.get(first).add(second);
            indegrees[second]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegrees[i] == 0) {
                queue.offer(i);
            }
        }

        int[] result = new int[numCourses];
        int index = 0;

        while (!queue.isEmpty()) {
            int cur = queue.poll();
            result[index] = cur;
            for (int nei : graph.get(cur)) {
                indegrees[nei]--;
                if (indegrees[nei] == 0) {
                    queue.offer(nei);
                }
            }
            index++;
        }

        if (index == numCourses) {
            return result;
        } else {
            return new int[0];
        }
    }

    public static void main(String[] args) {
        CourseScheduleII solution = new CourseScheduleII();

        int numCourses1 = 2;
        int[][] prerequisites1 = {{1, 0}};
        System.out.println(Arrays.toString(solution.findOrder(numCourses1, prerequisites1))); // Output: [0, 1]

        int numCourses2 = 4;
        int[][] prerequisites2 = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
        System.out.println(Arrays.toString(solution.findOrder(numCourses2, prerequisites2))); // Output: [0, 1, 2, 3]
    }
}

File: LeetCode_Python/graph/toplogical/CourseScheduleII.py
from collections import deque, defaultdict


class CourseScheduleII:
    def findOrder(self, numCourses: int, prerequisites: list[list[int]]) -> list[int]:
        graph = defaultdict(list)
        indegree = [0] * numCourses

        for pre in prerequisites:
            graph[pre[1]].append(pre[0])
            indegree[pre[0]] += 1

        queue = deque([i for i in range(numCourses) if indegree[i] == 0])

        res = []
        while queue:
            cur = queue.popleft()
            res.append(cur)
            for nei in graph[cur]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    queue.append(nei)

        return res if len(res) == numCourses else []


# Time Complexity: O(V + E), where V is the number of courses and E is the number of prerequisites.
# Space Complexity: O(V + E), for storing the graph and indegree array.

if __name__ == "__main__":
    solution = CourseScheduleII()

    numCourses1 = 4
    prerequisites1 = [[1, 0], [2, 0], [3, 1], [3, 2]]
    print("Expected Output: [0,1,2,3] or [0,2,1,3]")
    print("Actual Output:", solution.findOrder(numCourses1, prerequisites1))

    numCourses2 = 2
    prerequisites2 = [[1, 0]]
    print("Expected Output: [0,1]")
    print("Actual Output:", solution.findOrder(numCourses2, prerequisites2))

    numCourses3 = 2
    prerequisites3 = [[1, 0], [0, 1]]
    print("Expected Output: []")
    print("Actual Output:", solution.findOrder(numCourses3, prerequisites3))
